File: backend/api/app/data/repositories/post.py
from typing import Iterable
from core.common.either import Either
from core.errors.failure import Failure, CacheFailure
from core.errors.exceptions import CacheException
from app.domain.entities.post import Post, PostEntity
from app.domain.repositories.post import BaseRepository
from app.data.datasources.local.post import PostLocalDataSource

class PostRepositoryImpl(BaseRepository):
        
    def __init__(self, post_local_datasource: PostLocalDataSource):
        self.post_local_datasource = post_local_datasource
    
    async def create_post(self, post: Post) -> Either[Failure, PostEntity]:
        try:
            post_entity = await self.post_local_datasource.create_post(post)
            return Either.right(post_entity)
        except CacheException as e:
            return Either.left(CacheFailure(error_message=str(e)))
    
    async def update_post(self, post: Post) -> Either[Failure, PostEntity]:
        try:
            post_entity = await self.post_local_datasource.update_post(post)
            return Either.right(post_entity)
        except CacheException as e:
            return Either.left(CacheFailure(error_message=str(e)))
    
    async def delete_post(self, post_id: str) -> Either[Failure, PostEntity]:
        try:
            post_entity = await self.post_local_datasource.delete_post(post_id)
            return Either.right(post_entity)
        except CacheException as e:
            return Either.left(CacheFailure(error_message=str(e)))
    
    async def view_posts(self, user_id: str) -> Either[Failure, list]:
        try:
            posts = await self.post_local_datasource.view_posts(user_id)
            return Either.right(posts)
        except CacheException as e:
            return Either.left(CacheFailure(error_message=str(e)))
    
    async def view_post(self, post_id: str) -> Either[Failure, PostEntity]:
        try:
            post_entity = await self.post_local_datasource.view_post(post_id)
            return Either.right(post_entity)
        except CacheException as e:
            return Either.left(CacheFailure(error_message=str(e)))

    async def like_post(self, post_id: str, user_id: str) -> Either[Failure, PostEntity]:
        try:
            post_entity =  await self.post_local_datasource.like_post(post_id, user_id)
            return Either.right(post_entity)
        except CacheException as e:
            return Either.left(CacheFailure(error_message=str(e)))

    async def unlike_post(self, post_id: str, user_id: str) -> Either[Failure, PostEntity]:
        try:
            post_entity = await self.post_local_datasource.unlike_post(post_id, user_id)
            return Either.right(post_entity)
        except CacheException as e:
            return Either.left(CacheFailure(error_message=str(e)))

    async def clone_post(self, post_id: str, user_id: str) -> Either[Failure, PostEntity]:
        try:
            post_entity = await self.post_local_datasource.clone_post(post_id, user_id)
            return Either.right(post_entity)
        except CacheException as e:
            return Either.left(CacheFailure(error_message=str(e)))

    async def unclone_post(self, post_id: str, user_id: str) -> Either[Failure, PostEntity]:
        try:
            post_entity = await self.post_local_datasource.unclone_post(post_id, user_id)
            return Either.right(post_entity)
        except CacheException as e:
            return Either.left(CacheFailure(error_message=str(e)))
        
    async def all_posts(self, tags: list, search_word: str) -> Either[Failure, Iterable[PostEntity]]:
        try:
            posts = await self.post_local_datasource.all_posts(tags, search_word)
            return Either.right(posts)
        except CacheException as e:
            return Either.left(CacheFailure(error_message=str(e)))

File: mobile/lib/features/architect/data/repositories/post.dart
import 'package:dartz/dartz.dart';

import '../../../../core/errors/exception.dart';
import '../../../../core/errors/failure.dart';
import '../../../../core/network/network_info.dart';
import '../../domains/entities/post.dart';
import '../../domains/repositories/post.dart';
import '../datasources/local/auth.dart';
import '../datasources/remote/post.dart';

class PostRepositoryImpl extends PostRepository {
  PostRepositoryImpl({
    required this.remoteDataSource,
    required this.networkInfo,
    required this.authLocalDataSource,
  });

  final PostRemoteDataSource remoteDataSource;
  final NetworkInfo networkInfo;
  final AuthLocalDataSource authLocalDataSource;

  @override
  Future<Either<Failure, List<Post>>> all(
      {List<String>? tags, String? search}) async {
    if (await networkInfo.isConnected) {
      try {
        final auth = await authLocalDataSource.getToken();
        ;
        final posts =
            await remoteDataSource.allPosts(search, tags, auth.accessToken);
        return Right(posts);
      } on ServerException {
        return Left(ServerFailure());
      }
    } else {
      return Left(ServerFailure());
    }
  }

  @override
  Future<Either<Failure, Post>> clone(String id) async {
    if (await networkInfo.isConnected) {
      try {
        final auth = await authLocalDataSource.getToken();
        final post = await remoteDataSource.clonePost(id, auth.accessToken);
        return Right(post);
      } on ServerException {
        return Left(ServerFailure());
      }
    } else {
      return Left(ServerFailure());
    }
  }

  @override
  Future<Either<Failure, Post>> create(
      {required String image,
      required String title,
      String? content,
      required String userId,
      required List<String> tags}) async {
    print("create post");

    if (await networkInfo.isConnected) {
      try {
        final auth = await authLocalDataSource.getToken();
        final post = await remoteDataSource.createPost(
          image: image,
          title: title,
          content: content,
          tags: tags,
          userId: userId,
          token: auth.accessToken,
        );
        print("create post success");
        return Right(post);
      } on ServerException {
        print("create post fail");
        return Left(ServerFailure());
      }
    } else {
      print("create post fail");
      return Left(ServerFailure());
    }
  }

  @override
  Future<Either<Failure, Post>> delete(String id) async {
    if (await networkInfo.isConnected) {
      try {
        final auth = await authLocalDataSource.getToken();
        final post = await remoteDataSource.deletePost(id, auth.accessToken);
        print("delete post success");
        return Right(post);
      } on ServerException {
        return Left(ServerFailure());
      }
    } else {
      print("delete post fail");
      return Left(ServerFailure());
    }
  }

  @override
  Future<Either<Failure, Post>> like(String id) async {
    if (await networkInfo.isConnected) {
      try {
        final auth = await authLocalDataSource.getToken();
        final post = await remoteDataSource.likePost(id, auth.accessToken);
        return Right(post);
      } on ServerException {
        return Left(ServerFailure());
      }
    } else {
      return Left(ServerFailure());
    }
  }

  @override
  Future<Either<Failure, Post>> unclone(String id) async {
    if (await networkInfo.isConnected) {
      try {
        final auth = await authLocalDataSource.getToken();
        final post = await remoteDataSource.unLikePost(id, auth.accessToken);
        return Right(post);
      } on ServerException {
        return Left(ServerFailure());
      }
    } else {
      return Left(ServerFailure());
    }
  }

  @override
  Future<Either<Failure, Post>> unlike(String id) async {
    if (await networkInfo.isConnected) {
      try {
        final auth = await authLocalDataSource.getToken();
        final post = await remoteDataSource.unLikePost(id, auth.accessToken);
        return Right(post);
      } on ServerException {
        return Left(ServerFailure());
      }
    } else {
      return Left(ServerFailure());
    }
  }

  @override
  Future<Either<Failure, Post>> update(
      {required String id,
      required String image,
      required String title,
      String? content,
      required String userId,
      required List<String> tags}) async {
    if (await networkInfo.isConnected) {
      try {
        final auth = await authLocalDataSource.getToken();
        final post = await remoteDataSource.editPost(
          id: id,
          image: image,
          title: title,
          content: content,
          tags: tags,
          userId: userId,
          token: auth.accessToken,
        );
        print('update post success');
        return Right(post);
      } on ServerException {
        print('update post fail');
        return Left(ServerFailure());
      }
    } else {
      return Left(ServerFailure());
    }
  }

  @override
  Future<Either<Failure, Post>> view(String id) async {
    if (await networkInfo.isConnected) {
      try {
        final auth = await authLocalDataSource.getToken();
        final post = await remoteDataSource.viewPost(id, auth.accessToken);
        return Right(post);
      } on ServerException {
        return Left(ServerFailure());
      }
    } else {
      return Left(ServerFailure());
    }
  }

  @override
  Future<Either<Failure, List<Post>>> views(String userId) async {
    if (await networkInfo.isConnected) {
      try {
        final auth = await authLocalDataSource.getToken();
        final posts =
            await remoteDataSource.viewsPost(userId, auth.accessToken);
        return Right(posts);
      } on ServerException {
        return Left(ServerFailure());
      }
    } else {
      return Left(ServerFailure());
    }
  }
}
